### PR TITLE
fix isPlainObject

### LINKED ISSUE #304: Dispatching action from iframe to store defined in parent window always throws
Dispatching an action from an iframe to store defined in a parent window always throws `Invariant Violation: Actions must be plain objects. Use custom middleware for async actions.`

It fails on the check `Object.getPrototypeOf(obj) === Object.prototype` in `isPlainObject`.

Only way to fix this is to expose actions in the parent window.

### PR DESCRIPTION
**Pull request by michalkvasnicak**
When isPlainObject is called between two different contexts then Object.prototype is not the same. Now it is comparing only stringified constructors. Fixes #304

### CODE REVIEW

**Comment by gaearon** [comment]
Can you add a test that would have been failing with the previous version?

**Comment by michalkvasnicak** [comment]
I will try but don't know how to reproduce it inside the same context.

**Comment by gaearon** [comment]
Look into lodash tests? :-P

**Comment by phated** [comment]
Cc @jdalton

**Comment by michalkvasnicak** [comment]
It seems that lodash is creating its own context.

**Comment by michalkvasnicak** [comment]
OK I can reproduce it but had to install contextify for it. Can I push it with contextify as devDependency?

**Comment by gaearon** [comment]
Yes, sure.

File: src/utils/isPlainObject.js [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
  */
 export default function isPlainObject(obj) {
-  if (!obj) {
+  if (!obj || typeof obj !== 'object') {
     return false;
   }
 
-  return typeof obj === 'object' &&
-         Object.getPrototypeOf(obj) === Object.prototype;
+  const proto = typeof obj.constructor === 'function' ? Object.getPrototypeOf(obj) : Object.prototype;
```
Comment (gaearon) [inline src/utils/isPlainObject.js]: Please change `const` to `var`, we're maintaining Flow-friendliness for now. (I'll alter ESLint rule later.)

File: test/utils/isPlainObject.spec.js [hunk 2]
```diff
@@ -7,11 +8,17 @@ describe('isPlainObject', () => {
       this.prop = 1;
     }
 
+    const sandbox = contextify();
```
Comment (gaearon) [inline test/utils/isPlainObject.spec.js]: (It's fine to keep using `const` in tests)

**Comment by gaearon** [comment]
Looks good, thank you!

**Comment by michalkvasnicak** [comment]
I thank you for this awesome lib. :)

**Comment by jdalton** [comment]
fwiw lodash v4 drops IE8 support and has revised it's `isPlainObject` to be considerably more [light weight](https://github.com/lodash/lodash/blob/e7ad50d2434ff907ef266c20d870386a6c1a5ecf/lodash.js#L8250-L8264). You caught us at a time when we pulled our modern method fork of `isPlainObject` for a single larger implementation.

**Comment by jdalton** [comment]
It looks like this PR barrows from our WIP v4 implementation so you'll dig this method is designed to work with the es5-sham fallback for `Object.getPrototypeOf` too by punting if `constructor` is not a function.

**Comment by michalkvasnicak** [comment]
@jdalton Yes it is, I was pointed to look at the lodash implementation. I hope it's OK with you.

**Comment by gaearon** [comment]
@jdalton Appreciate you chiming in!

**Comment by jdalton** [comment]
@michalkvasnicak
>  I hope it's OK with you.

In the long term copy pasting is a bummer as it shifts the implementation testing burden to you all. It also means you all have to spend time mucking with and becoming experts in utils in stead of making cool stuff on top of it. It's on me to make the existing lodash tools more consumer friendly and file size is part of that. I hope this can be revisited in v4 where file size is a major focal point.

**Comment by gaearon** [comment]
@jdalton 

I agree totally. In our case we wanted to get some micro file size optimizations because the whole lib is 2KB minified+gzipped so we kind of sacrifice util maintainability. OTOH we only need three tiny utils so in my book it's worth the effort.

**Comment by jdalton** [comment]
Yap. As the old saying goes you can lead a horse to water but you sure as hell can't make them bust out a sweet jazz solo

**Pull request merged**